environ not declared on osx

I just noticed a bug compiling on OSX High Sierra. 
src/zproc.c:537:19: error: use of undeclared identifier 'environ'
            env = environ;

To be able to successfully compile you need to comment the else clause at line 537 of src/zproc.c then all is good. OSX does not provide environ support it seems.

Saw the same issue here; also using High Sierra. Was installing with brew:

```
==> ./configure --prefix=/usr/local/Cellar/czmq/HEAD-a6b87be
==> make
Last 15 lines from /Users/davidkarim/Library/Logs/Homebrew/czmq/03.make:
  CC       src/src_libczmq_la-zproxy.lo
  CC       src/src_libczmq_la-zrex.lo
  CC       src/src_libczmq_la-zgossip_msg.lo
  CC       src/src_libczmq_la-zargs.lo
  CC       src/src_libczmq_la-zproc.lo
  CC       src/src_libczmq_la-ztimerset.lo
  CC       src/src_libczmq_la-ztrie.lo
src/zproc.c:537:19: error: use of undeclared identifier 'environ'
            env = environ;
                  ^
1 error generated.
make[1]: *** [src/src_libczmq_la-zproc.lo] Error 1
make[1]: *** Waiting for unfinished jobs....
  CC       src/src_libczmq_la-czmq_private_selftest.lo
make: *** [all-recursive] Error 1
```

Would love to validate since I am working on High Sierra but I hadn't tried building it manually; before I used brew. Anyways, I did try pulling from vyskocilm:master and building manually but I got a different error. 
```
$ make check
Making check in doc
asciidoc -d manpage -b docbook -f ./asciidoc.conf \
		-azmq_version=4.2.4 -ozmq_bind.xml zmq_bind.txt
xmlto man zmq_bind.xml
xmlto: /Users/davidkarim/Documents/libzmq/doc/zmq_bind.xml does not validate (status 3)
xmlto: Fix document syntax or use --skip-validation option
I/O error : Attempt to load network entity http://www.oasis-open.org/docbook/xml/4.5/docbookx.dtd
/Users/davidkarim/Documents/libzmq/doc/zmq_bind.xml:2: warning: failed to load external entity "http://www.oasis-open.org/docbook/xml/4.5/docbookx.dtd"
D DocBook XML V4.5//EN" "http://www.oasis-open.org/docbook/xml/4.5/docbookx.dtd"
                                                                               ^
I/O error : Attempt to load network entity http://www.oasis-open.org/docbook/xml/4.5/docbookx.dtd
warning: failed to load external entity "http://www.oasis-open.org/docbook/xml/4.5/docbookx.dtd"
validity error : Could not load the external subset "http://www.oasis-open.org/docbook/xml/4.5/docbookx.dtd"
Document /Users/davidkarim/Documents/libzmq/doc/zmq_bind.xml does not validate
make[1]: *** [zmq_bind.3] Error 13
rm zmq_bind.xml
make: *** [check-recursive] Error 1
```

--- a/src/zproc.c
+++ b/src/zproc.c
@@ -60,7 +60,9 @@ content of the messages in any way. See test example on how to use it.
 
 #if defined (__UNIX__)
 #   if defined (__UTYPE_OSX)
-char **environ = *_NSGetEnviron (); // issue#1836
+// issue#1836
+#include <crt_externs.h>
+#define environ (*_NSGetEnviron ()) 
 #   else
 extern char **environ;              // should be declared as a part of unistd.h, but fail in some targets in Travis
                                     // declare it explicitly

--- a/src/zproc.c
+++ b/src/zproc.c
@@ -539,7 +539,7 @@ s_zproc_execve (zproc_t *self)
         }
         else
 #if defined (__UTYPE_OSX)
-            env = _NSGetEnviron ();
+            env = *_NSGetEnviron ();
 #else
             env = environ;
 #endif

--- a/src/zproc.c
+++ b/src/zproc.c
@@ -59,8 +59,11 @@ content of the messages in any way. See test example on how to use it.
 #include "czmq_classes.h"
 
 #if defined (__UNIX__)
-#   if !defined (_GNU_SOURCE)
-extern char **environ;          // declared in unistd.h if _GNU_SOURCE is defined
+#   if defined (__UTYPE_OSX)
+char **environ = *_NSGetEnviron (); // issue#1836
+#   else
+extern char **environ;              // should be declared as a part of unistd.h, but fail in some targets in Travis
+                                    // declare it explicitly
 #   endif
 #endif
 
@@ -544,11 +547,7 @@ s_zproc_execve (zproc_t *self)
             arr_add_ref (env, i, NULL);
         }
         else
-#if defined (__UTYPE_OSX)
-            env = *_NSGetEnviron ();
-#else
             env = environ;
-#endif
 
         r = execve (filename, argv2, env);
         if (r == -1) {

--- a/src/zproc.c
+++ b/src/zproc.c
@@ -65,6 +65,10 @@ content of the messages in any way. See test example on how to use it.
 # include <unistd.h>
 #endif
 
+#if defined (__UTYPE_OSX)
+#include <crt_externs.h>
+#endif
+
 #define ZPROC_RUNNING -42
 
 //      #######     internal helpers for zproc      #######
@@ -534,7 +538,11 @@ s_zproc_execve (zproc_t *self)
             arr_add_ref (env, i, NULL);
         }
         else
+#if defined (__UTYPE_OSX)
+            env = _NSGetEnviron ();
+#else
             env = environ;
+#endif
 
         r = execve (filename, argv2, env);
         if (r == -1) {

--- a/src/zproc.c
+++ b/src/zproc.c
@@ -58,6 +58,12 @@ content of the messages in any way. See test example on how to use it.
 
 #include "czmq_classes.h"
 
+#if defined (__UNIX__)
+#   if !defined (_GNU_SOURCE)
+extern char **environ;          // declared in unistd.h if _GNU_SOURCE is defined
+#   endif
+#endif
+
 // For getcwd() variants
 #if (defined (WIN32))
 # include <direct.h>

